client/mysqltest.cc

```diff
@@ -3865,26 +3865,268 @@ void do_remove_files_wildcard(struct st_command *command)
+{
+  int error= 0;
+
+  /* Note that my_dir sorts the list if not given any flags */
+  MY_DIR *src_dir_info= my_dir(ds_source->str,
+                               MYF(MY_DONT_SORT | MY_WANT_STAT));
+  MY_DIR *dst_dir_info= my_dir(ds_destination->str,
+                               MYF(MY_DONT_SORT | MY_WANT_STAT));
+
+  /* Either one of directories does not exist or access denied */
+  if (!src_dir_info || !dst_dir_info)
+  {
+    error= 1;
+    goto end;
+  }
+
+  char dir_separator[2];
+  dir_separator[0]= FN_LIBCHAR;
+  dir_separator[1]= 0;
+  dynstr_append_mem(ds_source, dir_separator, strlen(dir_separator));
+  dynstr_append_mem(ds_destination, dir_separator, strlen(dir_separator));
+
+  /* Storing the length of the path to the file, so it can be reused */
+  size_t source_file_length;
+  size_t dest_file_length;
+  dest_file_length= ds_destination->length;
+  source_file_length= ds_source->length;
+  uint match_count;
+  match_count= 0;
+
+  for (uint i= 0; i < src_dir_info->number_of_files; i++)
+  {
+    ds_source->length= source_file_length;
+    ds_destination->length= dest_file_length;
+    FILEINFO *file= src_dir_info->dir_entry + i;
+
+    /* Skip directories if recursive copy is not requested */
+    if (MY_S_ISDIR(file->mystat->st_mode) && !recursive)
+      continue;
+
+    /* Copy only those files which the pattern matches */
+    if (ds_wild->length &&
+        wild_compare(file->name, ds_wild->str, false))
+      continue;
+
+    match_count++;
+    dynstr_append_mem(ds_source, file->name, strlen(file->name));
+    dynstr_append_mem(ds_destination, file->name, strlen(file->name));
+
+    if (MY_S_ISDIR(file->mystat->st_mode))
+    {
+      DBUG_PRINT("info", ("Recursive copy files of %s to %s",
+                          ds_source->str, ds_destination->str));
+      DBUG_PRINT("info", ("listing directory: %s", ds_source->str));
+      error= recursive_copy(ds_source, ds_destination, do_not_overwrite, false);
+    }
+    else
+    {
+      DBUG_PRINT("info", ("Copying file: %s to %s",
+                          ds_source->str, ds_destination->str));
+      /* MY_HOLD_ORIGINAL_MODES prevents attempts to chown the file */
+      error= (my_copy(ds_source->str, ds_destination->str,
+              MYF(MY_HOLD_ORIGINAL_MODES)) != 0);
+    }
+
+    if (error) goto end;
+  }
+
+  /* Pattern did not match any files */
+  if (!match_count)
+  {
+    error= 1;
+  }
+
+end:
+  my_dirend(src_dir_info);
+  my_dirend(dst_dir_info);
+  return error;
+}
+
+
 /*
   SYNOPSIS
   do_copy_file
   command	command handle
 
   DESCRIPTION
-  copy_file <from_file> <to_file>
-  Copy <from_file> to <to_file>
+  copy_file <source> <destination> [-r|-f|-rf]
+
+  <source> can be a file, a directory, or a path with wildcard pattern
 
-  NOTE! Will fail if <to_file> exists
+  - When <source> is a directory, '-r' must be given to copy recursively
+  - When <source> is a wildcard pattern, '-r' is optional. If given, then
```

Comment:
a bit confusing. "-r is optional" suggests that with wildcards copying is always recursive. Or that without wildcards -r is required.
Neither seems to be the case, so better remove this item from the comment.